Check if a data-structure exists for hazelcast readers

gurbuzali opened the pull request:
Related: https://github.com/hazelcast/hazelcast-enterprise/issues/4257

hz-devops-test commented:
The job `Hazelcast-pr-builder` of your PR failed. (Hazelcast internal details: [build log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/6955/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/6955/Hazelcast-pr-builder-6955.zip)).
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   HazelcastConnectorTest.when_readCache_and_writeCache:288 expected:<100> but was:<0>
[ERROR] Errors: 
[ERROR]   JobSerializerTest.when_serializerIsRegistered_then_itIsAvailableForLocalCacheSource:135 ? Completion
[INFO] 
[ERROR] Tests run: 42801, Failures: 1, Errors: 1, Skipped: 1001
[INFO] 

[ERROR] There are test failures.
</pre>
</details>

hz-devops-test commented:
The job `Hazelcast-pr-builder` of your PR failed. (Hazelcast internal details: [build log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/6964/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/6964/Hazelcast-pr-builder-6964.zip)).
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   SqlBasicClientTest>SqlBasicTest.testSelect:289 expected:<4096> but was:<0>
[ERROR]   SqlBasicClientTest>SqlBasicTest.testSelect:289 expected:<4096> but was:<0>
[ERROR]   SqlBasicClientTest>SqlBasicTest.testSelect:178 expected:<4096> but was:<8192>
[ERROR]   SqlBasicClientTest>SqlBasicTest.testSelect:178 expected:<4096> but was:<8192>
[ERROR]   SqlBasicClientTest>SqlBasicTest.testSelect:178 expected:<4096> but was:<12288>
[ERROR]   SqlBasicClientTest>SqlBasicTest.testSelect:178 expected:<4096> but was:<12288>
[INFO] 
[ERROR] Tests run: 3626, Failures: 6, Errors: 0, Skipped: 63
[INFO] 

[ERROR] There are test failures.
</pre>
</details>

mmedenjak commented:
@gurbuzali @frant-hartm can we make this PR ready for review and merged?

Commit:
Check if a data-structure exists for hazelcast readers
Check if a data-structure exists for secured functions
Remove CREATE action from permissionsMemberDomConfigProcessor.java:3017

gurbuzali commented:
@mmedenjak I've rebased it against master, the PR should be ready for review.
The main thing we should discuss is that we are relying on the proxy-service for checking if a data-structure exists.

hz-devops-test commented:
The job `Hazelcast-pr-builder` of your PR failed. (Hazelcast internal details: [build log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/7621/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/7621/Hazelcast-pr-builder-7621.zip)).
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   ReliableTopicDestroyTest.whenDestroyedThenRingbufferRemoved:88->HazelcastTestSupport.assertTrueEventually:1362->HazelcastTestSupport.assertTrueEventually:1260
[INFO] 
[ERROR] Tests run: 42950, Failures: 1, Errors: 0, Skipped: 1011
[INFO] 

[ERROR] There are test failures.
</pre>
</details>

frant-hartm commented on hazelcast/src/test/java/com/hazelcast/jet/impl/connector/HazelcastConnectorTest.java:
```diff
@@ -106,6 +109,46 @@ public void before() {
         cacheManager.getCache(streamSinkName);
     }
 
+    public void when_readNonExistedMap_then_mapNotCreated() {
```
```suggestion
    public void when_readNonExistingMap_then_mapNotCreated() {
```

frant-hartm commented on hazelcast/src/test/java/com/hazelcast/jet/impl/connector/HazelcastConnectorTest.java:
```diff
@@ -106,6 +109,46 @@ public void before() {
+        assertTrue(instance().getDistributedObjects().stream().noneMatch(d -> d.getName().equals(sourceName)));
+    }
+
+    public void when_queryNonExistedMap_then_mapNotCreated() {
```
```suggestion
    public void when_queryNonExistingMap_then_mapNotCreated() {
```

frant-hartm commented on hazelcast/src/test/java/com/hazelcast/jet/impl/connector/HazelcastConnectorTest.java:
```diff
@@ -106,6 +109,46 @@ public void before() {
+
+        instance().getJet().newJob(p).join();
+
+        assertTrue(instance().getDistributedObjects().stream().noneMatch(d -> d.getName().equals(sourceName)));
```
We could use better assert here, e.g.

```
assertThat(instance.getDistributedObjects())
                .extracting(DistributedObject::getName)
                .doesNotContain(sourceName);
```

It would give a better error message in case of failure and is easier to read.

frant-hartm replied:
Also in other places where this is used

frant-hartm commented on hazelcast/src/test/java/com/hazelcast/jet/impl/connector/HazelcastConnectorTest.java:
```diff
@@ -106,6 +109,46 @@ public void before() {
+        assertTrue(instance().getDistributedObjects().stream().noneMatch(d -> d.getName().equals(sourceName)));
+    }
+
+    public void when_readNonExistedCache_then_cacheNotCreated() {
```
```suggestion
    public void when_readNonExistingCache_then_cacheNotCreated() {
```

frant-hartm commented on hazelcast/src/test/java/com/hazelcast/jet/impl/connector/HazelcastConnectorTest.java:
```diff
@@ -106,6 +109,46 @@ public void before() {
+        assertTrue(instance().getDistributedObjects().stream().noneMatch(d -> d.getName().endsWith(sourceName)));
+    }
+
+    public void when_readNonExistedList_then_listNotCreated() {
```
```suggestion
    public void when_readNonExistingList_then_listNotCreated() {
```

frant-hartm reviewed (changes requested):
Some minor comments, but most importantly I don't see anything for remote sources - remoteMap/Cache/List - shouldn't these be handled as well?

Commit:
address review comments

gurbuzali commented:
@frant-hartm  the reasoning of this PR heavily depends on the security. Currently you need to have a `create` and `read` permission in order to use a map as a source. With this PR, a jet job won't create a map if it is not created already thus we'll be able to remove the `create` permission from the required permission set.

For remote-sources, the check is not cheap. You need to call `getDistributedObjects` which will fetch all the data-structure info from a member. Or we can introduce a new client-message which will just check if a DS is created or not instead of fetching them all. In any case it shall be done to make it consistent.

Commit:
add check for remote sources

devOpsHazelcast commented:
[![CLA assistant check](https://cla.hazelcast.org:443/pull/badge/not_signed)](https://cla.hazelcast.org:443/hazelcast/hazelcast?pullRequest=19440) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.hazelcast.org:443/hazelcast/hazelcast?pullRequest=19440) before we can accept your contribution.<br/><hr/>**Ali Gurbuz** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.hazelcast.org:443/check/hazelcast/hazelcast?pullRequest=19440) it.</sub>

viliam-durina commented:
How to proceed with this PR? I'd close it unless there's a strong reason to stop requiring the CREATE permission for map reading. HZ is designed that objects are auto-created, if we want to change this, we should go from the internal design (possibly in hz 6). Otherwise we just create confusion and technical debt.

devOpsHazelcast commented:
Can one of the admins verify this patch?

devOpsHazelcast commented:
Can one of the admins verify this patch?

devOpsHazelcast commented:
Can one of the admins verify this patch?
